fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1689,23 +1771,31 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
         # reductions in post_backward stream.
         self._streams["post_backward"].wait_stream(torch.cuda.current_stream())
         with torch.cuda.stream(self._streams["post_backward"]):
-            orig_grad_data = param.grad.data
+            orig_grad_data = grad_or_main_grad
 
-            if self.mixed_precision and self.fp32_reduce_scatter:
-                # Cast grad to FP32.
-                param.grad.data = param.grad.data.to(param.dtype)
+            if self.mixed_precision:
+                if self.fp32_reduce_scatter:
+                    # Cast grad to FP32.
+                    grad_or_main_grad.data = grad_or_main_grad.to(param.dtype)
+                elif self._is_fp8_dtype():
+                    # Use bf16 wgrad for fp8 weights (TODO: handle fp8 wgrad)
```

Comment:
Currently this is not working with the latest FP8 wgrad ?

Reply:
This meant to be for future work when we have fp8 reduce-scatter. I'll update the comment.